File: clustering/airlines.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jul 11 18:51:48 2020

@author: Ranna
"""
import pandas as pd
import matplotlib.pyplot as plt 
import numpy as np
airline = pd.read_excel("C:\\Users\\Ranna\\Documents\\excelR\\assignments\\clustering\\EastWestAirlines.xlsx", sheet_name=1)

airline.info()
airline.describe()

# 1st moment
airline.mean()
airline.mode()
airline.median()

# 2nd moment
# sample var & std
airline.var()
airline.std()
# populdtion var & std
np.var(airline)
np.std(airline)

# 3rd & 4th moments
airline.skew()
airline.kurt()

########## H-clustering ################

## data has many ranges, hence normalization is necessary
# Normalization function 
def normal(i):
    x = (i-i.min())/(i.max()-i.min())
    return (x)

# Normalized data frame 
df_norm = normal(airline.iloc[:,1:]) #(considering the numerical part of data)
df_norm.describe()

from scipy.cluster.hierarchy import linkage 
import scipy.cluster.hierarchy as sch # for creating dendrogram 


z = linkage(df_norm, method="complete",metric="euclidean")

plt.figure(figsize=(15, 5));plt.title('Hierarchical Clustering Dendrogram');plt.xlabel('Index');plt.ylabel('Distance')
sch.dendrogram(
    z,
    leaf_rotation=0.,  # rotates the x axis labels
    leaf_font_size=8.,  # font size for the x axis labels
)
plt.show()

# Now applying AgglomerativeClustering choosing 3 as clusters from the dendrogram
from sklearn.cluster import	AgglomerativeClustering 
h_complete	=	AgglomerativeClustering(n_clusters=6,linkage='complete',affinity = "euclidean").fit(df_norm) 

h_complete.labels_

cluster_labels=pd.Series(h_complete.labels_)

airline['clust']=cluster_labels # creating a  new column and assigning it to new column 
airline = airline.iloc[:,[1,2,3,4,5,6,7,8,9,10,11,12]]
airline.head()

# getting aggregate mean of each cluster
airline.groupby(airline.clust).mean()

# creating a csv file 
airline.to_csv("airline_dendo.csv",index=False) #,encoding="utf-8")


############## k-means ####################
from sklearn.cluster import	KMeans
from scipy.spatial.distance import cdist 

k = list(range(2,15))
k
TWSS = [] # variable for storing total within sum of squares for each kmeans 
for i in k:
    kmeans = KMeans(n_clusters = i)
    kmeans.fit(df_norm)
    WSS = [] # variable for storing within sum of squares for each cluster 
    for j in range(i):
        WSS.append(sum(cdist(df_norm.iloc[kmeans.labels_==j,:],kmeans.cluster_centers_[j].reshape(1,df_norm.shape[1]),"euclidean")))
    TWSS.append(sum(WSS))

# Scree plot 
plt.plot(k,TWSS, 'ro-');plt.xlabel("No_of_Clusters");plt.ylabel("total_within_SS")


from sklearn.metrics import silhouette_score 

for n_clusters in k:
    clusterer = KMeans(n_clusters=n_clusters)
    preds = clusterer.fit_predict(df_norm)
    centers = clusterer.cluster_centers_

    score = silhouette_score(df_norm, preds)
    print("For n_clusters = {}, silhouette score is {})".format(n_clusters, score))

plt.plot(n_clusters, score)
plt.title("Silhouette score values vs Numbers of Clusters ")
plt.show()

# Selecting 5 clusters from the above scree plot which is the optimum number of clusters 
model=KMeans(n_clusters=5) 
model.fit(df_norm)

model.labels_ # getting the labels of clusters assigned to each row 
md=pd.Series(model.labels_)  # converting numpy array into pandas series object 
airline['clust']=md
airline.head(10) # creating a  new column and assigning it to new column 
airline = airline.iloc[:,[2,3,4,5,6,7,8,9,10,11,12]]
airline.iloc[:,1:7].groupby(airline.clust).mean()

File: clustering/airlines.R
library(readxl)
airlines <- read_excel("C:/Users/Ranna/Documents/excelR/assignments/clustering/EastWestAirlines.xlsx", sheet = "data")
View(airlines)

#EDA
str(airlines)
summary(airlines)
library(skimr)
skim(airlines)
plot(airlines)
boxplot(airlines)

#Heirarchial_clustering
airlines1 <- airlines[,2:12]
View(airlines1)
norm_airlines1 <- scale(airlines1)
View(norm_airlines1)
d <- dist(norm_airlines1, method = "euclidean")
fit <- hclust(d, method = "complete")
fit1 <- hclust(d, method = "ward.D2")
plot(fit, hang = -1)
plot(fit1, hang = -1)
rect.hclust(fit1, k=12, border="red")

groups <- cutree(fit, k=) # cut tree into 4 clusters

membership<-as.matrix(groups) # groups or cluster numbers
final <- data.frame(airlines, membership)

View(final)

write.csv(final, file="final_airlines.csv",row.names = F)

aggregate(airlines[,-1],by=list(final$membership),mean)

########################
#k-means clustering

wssplot <- function(norm_airlines1,nc=15,seed=123){
  wss <- (nrow(norm_airlines1)-1)*sum(apply(norm_airlines1,2,var))
  
  for (i in 2:nc) {
    set.seed(seed)
    wss[i] <- sum(kmeans(norm_airlines1,centers = i)$withinss)
  }
  #windows()
  plot(1:nc,wss,type = "b",xlab = "Number of clusters",ylab = "Within groups sum of squares")
}

wssplot(norm_airlines1)
########## elbow plot shows 10 number of clusters
#k selection method
library(kselection)
#install.packages("doParallel")
library(doParallel)
registerDoParallel(cores = 4)
k <- kselection(norm_airlines1,parallel = TRUE, k_threshold = 0.85)  
k
########k selection method also shows 10 clusters
######################################
#install.packages("NbClust")
#library(NbClust)
#set.seed(123)
#devAskNewPage(ask = T)
#nc <- NbClust(norm_airlines1,min.nc = 2,max.nc = 12, method="kmeans")
#table(nc$Best.n[1,])
#barplot(table(nc$Best.n[1,]),xlab = "Number of clusters",ylab="Number of Criteria",
#        main = "number of clusters chosen by 26 criteria")
#set.seed(123)
#fit.km <- kmeans(norm_airlines1,3,nstart = 2)
#fit.km$size
#fit.km$centers
#aggregate(airlines1,by=list(cluster=fit.km$cluster),mean)
#ct.km <- table(airlines$Balance,fit.km$cluster)
#ct.km
#install.packages("flexclust")
#library(flexclust)
#randIndex(ct.km)

####above code not working#####
#####################################
#library(animation)

#km<-kmeans.ani(norm_airlines1,3)
#km$cluster
#km$withinss
#km$centers

#partitioning around medoids
library(cluster)
set.seed(1234)
fit.pam <- pam(norm_airlines1,k=10,stand = TRUE)
fit.pam$medoids
clusplot(fit.pam, main = "Bivariate Cluster Plot")
#ct.pam <- table(airlines1,fit.pam$clustering)
#ct.pam
#randIndex(ct.pam)

# # k clustering alternative for large dataset - Clustering Large Applications (Clara)
# install.packages("cluster")
library(cluster)
xcl <- clara(norm_airlines1, 10)
clusplot_model <- clusplot(xcl)
str(xcl)


# the medoids and clara model both show the variability at 45.58%, so 10 number of  clusters should be right

